### ISSUE TITLE
How to configure to use shared audio

### ISSUE THREAD

**Issue by peteh** [GioF71/squeezelite-docker#55]
It seems like per default the docker player uses exclusive access of the sound device. 

Problem: 
 * Starting the docker container and playing music works
 * Trying to start a video on youtube in parallel does not work
If I turn off the player the youtube video starts. 

I'm running kubuntu 22.04 and start the player with the following docker-compose: 
```
version: "3"

services:
  squeezelite-super-amazing-dac:
    image: giof71/squeezelite:latest
    container_name: squeezelite-super-amazing-dac
    devices:
      - /dev/snd:/dev/snd
    environment:
      - PRESET=super-amazing-dac,archimago-goldilocks,rates_up_to_384k
      - SQUEEZELITE_NAME=super-amazing-dac
      - SQUEEZELITE_SERVER_PORT=192.168.2.50
      - DISPLAY_PRESETS=Y
      - SQUEEZELITE_NAME="Living Room Computer"
      - SQUEEZELITE_AUDIO_DEVICE=default
```

I assumed that default would use pulse which is supposed to multiplex the audio streams. Where am I going wrong in my setup? 

Error messages when trying to enable the audio player from LMS while already playing a youtube video: 
```
squeezelite-super-amazing-dac  | Command Line: [/usr/bin/squeezelite -n "Living Room Computer" -C 2 -D 500 -s 192.168.2.50 -r 44100-384000:200 -R v::4:28:95:105:45]
squeezelite-super-amazing-dac  | [14:17:25.957504] alsa_open:360 playback open error: Device or resource busy
squeezelite-super-amazing-dac  | [14:17:30.966192] alsa_open:360 playback open error: Device or resource busy
squeezelite-super-amazing-dac  | [14:17:35.966741] alsa_open:360 playback open error: Device or resource busy
```

**Comment by GioF71** [comment]
Hello Pete,
SqueezeLite uses ALSA in this container image and in alsa mode SqueezeLite uses the audio device exclusively. This avoid resampling issues and yields a good audio quality. You should use that mode (e.g.) for you dedicated audio players connected to a DAC to your amplifier.

On the other end, it's not a good idea to use this mode on your desktop computer because you will encounter this issue.
You should use squeezelite with the PulseAudio backend. 
You can do that very easily using this repo: (https://github.com/GioF71/squeezelite-pulse-user-service). This does exactly what you need: you will also be able to select the output for the squeezelite player and to control its volume. However, if you play music and then start a youtube video, thing will start to mix and this is not ideal if you are looking for a good quality music reproduction.
Also, all of your music will generally be resampled (when necessary) to 44.1kHz (or 48.0). This is mandatory AFAIK for shared audio, because a device can play one specific sample rate at the same time. 

Of course, if your computer has multiple audio outputs, you actually can use this container image and let the container use that dac. In this case, it would be better to tell PulseAudio to not use that device.

Let me know if this helps.


Cheers
Giovanni

**Comment by GioF71** [comment]
Ah, btw, your compose file is using the default audio output via ALSA. 
Can you share your setup?
Thanks

**Comment by peteh** [comment]
Hi, thanks for the quick response. 

Here is my setup: 

Device A (Living Room): 
 * Raspberry pi 4
 * Rhasspy (voice assistant that also talks back)
 * Home assistant
 * Node-Red
 * Logitech Media Server instance
 * All services are running in docker with docker-compose files
 * Plan: use as squeezebox client to play music as well from LMS

Device B (Office): 
 * Raspberry Pi 4
 * Rhasspy sattelite for Device A (forwards audio to device a for recognition, plays text to speech from device A)
 * Plan: use a squeezebox client to play music as well from LMS

Device C (Living room): 
 * My main computer
 * Uses Kubuntu 22.04
 * Plan: use as squeezebox client to play music as well from LMS

The problem is that rhasspy and squeezebox would have to share audio or I need to have two sound outputs and separate speakers which is inconvenient. 

I'm still learning sound devices in linux which seems to be a more complex topic than on Windows. I would like to use docker still as it makes it super easy to setup and I can easily clone the setup on multiple devices with similar configs. 

What I understand is that I'd have to install squeezebox directly on the host to make use of the pulse backend and I probably would have to solve access of rhasspy to the sound card as well (I think it currently also uses alsa):
![image](https://user-images.githubusercontent.com/918728/188284477-cb4d165d-c0d6-4721-bfb9-9f329c4af1ad.png)

Thus to make both work on the same device they both would have to be mixed using pulse audio. 

Additionally to make the Main computer work as squeezebox client, the easiest seems to be to also install on the host and just make it use pulse audio. 

I also plan to add more satellites in the setup based on pi zero2w, that should also work as squeeze clients and rhasspy sattelites at the same time without duplicating the sound setup.

**Comment by GioF71** [comment]
Hello @peteh 
PulseAudio with docker is possible, although not without issues. It needs the user to be logged on to work properly (at least AFAIK) so I ended up controlling it via a user systemd service instead of the "regular" docker restart policy.
See my repo: https://github.com/GioF71/mpd-pulse-docker/
Maybe something similar can be done for squeezelite, but again, AFAIK, it is tricky and it's hard to set up properly in a headless system. Maybe a hidden vnc session with a desktop environment could do the trick, but would waste a lot of resources.

Maybe using an audio mixer downstream would solve the issue easy, letting you use different audio devices?
Something like this product

https://www.amazon.it/portatile-canali-Studio-stereo-passivo/dp/B07Y32K1FP/ref=sr_1_24?__mk_it_IT=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=10B3D7C2IGDBO&keywords=2+channel+mixer+audio&qid=1662239947&sprefix=2+channel+mixer+audio%2Caps%2C102&sr=8-24

or on amazon.com:

https://www.amazon.com/Portable-Channel-Audio-Passive-Recording/dp/B07Y2RYHFF

I know it's not ideal but it's quite cheap and might help keep things simple.

Let me know what you think!

**Comment by peteh** [comment]
I also thought about combining the 2 audio streams on hardware level. If the volume doesn't need to be adjusted I think it also just works to have two female plugs and wire them to a male output plug and use as an adapter. 
However it would require two sound devices which at least for the pi zero setup would make it less compact. 

For rhasspy somebody already built a docker setup that uses pulse which I probably will try at one of the next few weekends. 

The logging-in problem is not such a big deal as I'm using a DE anyway on the PIs, it automatically logs in and therefore will probably work. 

I will waste some time in figuring pulse out with your docker setup and the one from rhasspy. If I can't get it working I'll maybe fall back to the hardware solution or reach out here again if I'm stuck at a certain point. 

I will close here for now as I think the general question is answered for the current state. 

Thank you very much for your answers and your time! It helped to understand the whole situation and how to proceed a lot.

**Comment by GioF71** [comment]
Hello @peteh, would you like to try the latest release?
I have introduced experimental support for pulseaudio and hopefully didn't break alsa mode. 
https://github.com/GioF71/squeezelite-docker/releases/tag/2022-09-15-pulse

**Comment by peteh** [comment]
Thanks @GioF71, works like a charm on my living room Kubuntu 22.04 Ryzen machine. Thank you for putting in the effort to make this work. 

You wrote that it has some issue with rebooting when using restart policy. Do you know why? Is it because the devices are not ready? 

Potential ideas to avoid the workaround with the systemd service: 
 * Would it be possible to make a configurable delay in the container before starting the squeezebox binary to avoid this issue? Like a configurable sleep via environment env?
 * Can we see if the device is not ready yet to not start yet and query until it's ready if PULSE mode is used?

**Comment by GioF71** [comment]
Hello @peteh, it is nice to hear it worked for you!
I am now using this solutiont too, instead of the other one I suggested you (this one: https://github.com/GioF71/squeezelite-pulse-user-service)

There is already a DELAY parameter (SQUEEZELITE_DELAY), in seconds, that you might want to use. But the problem is that PulseAudio is not available until the Login is complete, afaik. If the computer needs a human interaction for login, it is tricky to set a good delay that would not be unbearably long.
The issues: audio not working until a proper restart. That lead me to understand the issue had to do with the automatic restart of the container.

About this question:

_Can we see if the device is not ready yet to not start yet and query until it's ready if PULSE mode is used?_

it seems a good idea but at the moment I do not know how it might be possible to implement it. I will try to understand if there is something I can do.

Of course, any input is very well welcome.

Cheers
Giovanni

**Comment by peteh** [comment]
I still found some issues: 
Raspberry 4 setup with internal headphone jack and additional usb audio: 
 * when using the usb audio as default output it just stutters and I can't get it to work
 * when using the internal headphone jack as default output in raspi-config it works

Raspberry pi 4 just using headphone jack: 
 * so far having experienced any issues in this setup

Normal PC with ryzen cpu and asus xonar dx sound card
 * working fine until I change volume with the volume keys on the keyboard - then it stutters as in the first raspberry setup

All other solutions from other  projects I've seen also map the pulse cookie in $HOME/.config/pulse/cookie
E.g. my rhasspy docker compose file looks like this: 
```
version: '3.3'
services:
    rhasspy:
        ports:
            - '12101:12101'
        container_name: rhasspy
        restart: unless-stopped
        volumes:
            - ${HOME}/.home/rhasspy/profiles:/profiles
            - /etc/localtime:/etc/localtime:ro
            - /run/user/1000/pulse:/run/user/1000/pulse
            - ${HOME}/.config/pulse/cookie:/home/pacat/.config/pulse/cookie
        devices:
            - /dev/snd:/dev/snd
        image: rhasspy/rhasspy
        command: --profile en --user-profiles /profiles
```

I don't have very detailed information about pulse, but I guess the other solutions take the cookie somehow into account. It requires you to run the container from your local user instead and have it added to the docker group I assume.

**Comment by GioF71** [comment]
Try running with --user and to map the cookie as well. Although I don't believe this might have something to do with stuttering. Does the usb dac get enough power?
However, I was thinking about using user mode in docker, so i will definitely try, probably tomorrow.

**Comment by GioF71** [comment]
Hello, have a look at this part of the code:
https://github.com/GioF71/squeezelite-docker/blob/main/app/bin/run-squeezelite-pulse.sh
it should alread run squeezelite as the specified user. I have also checked on my laptop, and the process looks like this:

giovanni    2698  4.1  0.4 313924 34160 ?        SLsl set18   1:20 /usr/bin/squeezelite -o pulse -s lms.homelab.com -n acer-aspire-5750g-pulse

Also I would like to add that this is working properly with onboard analog audio and is working properly on my desktop with onboard digital output. I can also control the volume from the LMS web interface.
I did not have to mount the pulse cookie, however. My operating system is Manjaro in both case (although on the laptop it's KDE and it's Gnome on the desktop).
Did you try with the cookie?

Also, check with `cat /proc/asound/cards` to see if your DAC is running in "high-speed" or "full-speed" mode. Sometime it happens that some DAC negotiate lower rates and this has resulted in issues for me in the past. I don't know if this is your problem, but it's maybe worth checking.

Ah, I just build a new bugfix release, I broke some default values leading to non-functional stacks in alsa mode due to an invalid command line. 
This should be fixed now.

**Comment by peteh** [comment]
@GioF71 I haven't tried with the cookie as I'm not exactly sure how to mount it properly into the container. 
What I can tell: 
when running on Kubuntu 22.04 on my ryzen machine: 
 - starting the container works
 - playing music on it works too and I can change the volume from the server
 - When I change the volume on the ryzen machine (in the actual desktop of it with the volume buttons - not from the server) it starts stuttering and also restarting the container will result in it stuttering until I fully restart the machine
 - stopping the container also stops the stuttering

Should I maybe create a new issue for it?

**Comment by GioF71** [comment]
Hello @peteh, yes please open another ticket. Maybe add the full docker-compose file as well.
The audio set might also be useful.
So maybe the output of those:

`cat /proc/asound/cards`

`aplay -l`
`aplay -L`

Thank you

**Comment by peteh** [comment]
@peteh done, so we can leave this issue here
